virtcontainers/device/api/interface.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2017-2018 Intel Corporation
+// Copyright (c) 2018 Huawei Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package api
+
+import (
+	"github.com/sirupsen/logrus"
+
+	"github.com/kata-containers/runtime/virtcontainers/device/config"
+)
+
+var devLogger = logrus.FieldLogger(logrus.New())
+
+// SetLogger sets the logger for virtcontainers package.
+func SetLogger(logger logrus.FieldLogger) {
+	devLogger = logger
+}
+
+// DeviceLogger returns logger for device management
+func DeviceLogger() *logrus.Entry {
+	return devLogger.WithField("subsystem", "device")
+}
+
+// DeviceReceiver is an interface used for accepting devices
+// a device should be attached/added/plugged to a DeviceReceiver
+type DeviceReceiver interface {
```

Comment:
@WeiZhang555 I'm trying to understand why you need Sandbox to implement these interfaces. Is it because we do not export proper hypervisor APIs? The layering looks a bit strange but if it is temporary until we have a hypervisor module, I'm ok with it.

Reply:
Yes, this is temporary, if we can make a good hypervisor module things can be easier.
We need more refactor including seperating hypervisor management part. 
And existence of `GetAndSetSandboxBlockIndex()` and `DecrementSandboxBlockIndex()` means we have to use Sandbox as `DeviceReceiver`. But actually a `Hypervisor` will be a better `DeviceReceiver`.

The implementation is not complete and perfect, we need more furthur works. But I don't have enough time to do more.

Reply:
Agree, this interface currently looks strange and forced, combining the calls for Sandbox and Hypervisor.
@WeiZhang555 I understand that you are trying to decouple devices completely from the rest of the virtcontainers components, but I would like to understand more clearly why this is required and how this will help the storage hotplug case. How storage hotplug will look without this decoupling.

Like you pointed out, devices do depend on calls from the sandbox and the hypervisor as well. I am not sure if we can get way with decoupling from the hypervisor completely, as we may have to call into some hypervisor specific details.

Reply:
@amshinde devices should be able to only depend on hypervisor, once we have a hypervisor module there. It calls from the sandbox only because hypervisor is part of sandbox right now. There are two dependences on sandbox, hotplug api and virtio-scsi device index management. The hotplug api is just a wrapper of hypervisor hotplug api. And the virtio-scsi device index should be just put inside the hypervisor as well because it is hypervisor specific data.

If we all agree this code needs refactoring/decoupling, it looks to be a reasonable step towards a proper shape of modular components in virtcontainers.

Reply:
Thanks @bergwolf for answering this for me! Your said exactly what I want to say!

The problem I met while doing the storage hotplug feature is that the devices, sandbox, hypervisor and containers are all coupled closely. What I only want to do is add a "device" to "hypervisor", while at this time I don't even need a container, so first thing I want to do is remove "container" from the device manager API.

This PR is trying to separate the "hotplug device to hypervisor" part, but as @bergwolf said, the block index is put in sandbox  structure, so I have to use both sandbox and sandbox.hypervisor, that's why the sandbox is a "DeviceReceiver" but not the `hypervisor`, in my head `hypervisor` could a good enough `DeviceReceiver`, that can be done in future and needs more modifications.

@amshinde

Reply:
@bergwolf @WeiZhang555 Thanks for the explanation, it makes sense to me now. I agree we should move the virtio-scsi index to the hypervisor. 
@WeiZhang555 Can we create an issue for this, so that we can keep track.